pkg/network/ovn_kubernetes.go

```diff
@@ -1526,5 +1522,64 @@ func isInternalSubnetLargeEnough(conf *operv1.NetworkSpec, v4 bool) bool {
 	}
 	// We need to ensure each node can be assigned an IP address from the internal subnet
 	intSubnetMask, _ := strconv.Atoi(strings.Split(subnet, "/")[1])
-	return maxNodesNum < (1<<(addrLen-intSubnetMask) - 2)
+	// reserve one IP for the gw, one IP for network and one for broadcasting
+	return maxNodesNum < (1<<(addrLen-intSubnetMask) - 3)
+}
+
+func isV6InternalSubnetLargeEnough(conf *operv1.NetworkSpec) bool {
+	var maxNodesNum uint128
+	var addrLen uint32
+	subnet := conf.DefaultNetwork.OVNKubernetesConfig.V6InternalSubnet
+	addrLen = 128
+	for _, n := range conf.ClusterNetwork {
+		if !utilnet.IsIPv6CIDRString(n.CIDR) {
+			continue
+		}
+		mask, _ := strconv.Atoi(strings.Split(n.CIDR, "/")[1])
+		nodesNum := &uint128{lo: 1}
+		nodesNum.shiftLeft(n.HostPrefix - uint32(mask))
+		maxNodesNum.add(nodesNum)
+	}
+	// We need to ensure each node can be assigned an IP address from the internal subnet
+	intSubnetMask, _ := strconv.Atoi(strings.Split(subnet, "/")[1])
+	capacity := &uint128{lo: 1}
+	capacity.shiftLeft(addrLen - uint32(intSubnetMask))
+	// reserve one IP for the gw, one IP for network and one for broadcasting
+	return maxNodesNum.add(&uint128{lo: 3}).isSmallerThan(capacity)
+}
+
+type uint128 struct {
+	hi, lo uint64
+}
+
+func (i *uint128) isSmallerThan(n *uint128) bool {
+	if i.hi != n.hi {
+		return i.hi < n.hi
+	} else {
+		return i.lo < n.lo
+	}
+}
+
+func (i *uint128) add(n *uint128) *uint128 {
+	origlow := i.lo
+	i.lo += n.lo
+	i.hi += n.hi
+	if i.lo < origlow { // wrapping occurred, so carry the 1
+		i.hi += 1
+	}
+	return i
+}
+
+func (i *uint128) shiftLeft(bits uint32) {
+	if bits >= 128 {
+		i.hi = 0
+		i.lo = 0
+	} else if bits >= 64 {
+		i.hi = i.lo << (bits - 64)
+		i.lo = 0
+	} else {
+		i.hi <<= bits
+		i.hi |= i.lo >> (64 - bits)
+		i.lo <<= bits
+	}
```

Comment:
Could we place this in a reusable place with some tests?
Any reason not to use `Int` from `math/big`?

Reply:
I was not aware it exists. Let me refactor the patch with `math/big`. Thanks.